**Практична робота: моделювання в озері даних**

In [30]:
pip install pandas pyarrow azure-storage-blob

Note: you may need to restart the kernel to use updated packages.


In [31]:
from azure.storage.blob import ContainerClient
from io import BytesIO
import pandas as pd

**Завдання 1. Підключення**

**Завдання 1.1. Створіть ноутбук medallion.ipynb, вставте комірку підключення і замініть STUDENT на власний префікс формату prizvysche_imya латиницею, малими літерами, без пробілів.**

In [32]:
# 1. Вставьте ваши полные SAS-ссылки из предыдущей практической работы
# Джерело (тільки читання)
SOURCE_SAS = "https://datalakedatalab.blob.core.windows.net/data-lake?sp=r&st=2026-07-16T08:11:59Z&se=2026-08-31T16:26:59Z&spr=https&sv=2026-02-06&sr=c&sig=JHmTjds1p8Rp%2Bn9nXZUSWSMdkgAkuPdz%2B4KjbAQUlX0%3D"

# Озеро (читання + запис)
LAKE_SAS = "https://datalakedatalab.blob.core.windows.net/students-homework?sp=racwl&st=2026-07-16T08:10:00Z&se=2026-08-31T16:25:00Z&spr=https&sv=2026-02-06&sr=c&sig=wCRv3poEsB8L4%2Bnj3BKZvwcvlX9vGJ5xo93jYaG3no4%3D"

source = ContainerClient.from_container_url(SOURCE_SAS)
lake   = ContainerClient.from_container_url(LAKE_SAS)

STUDENT = "olkhovets_olena"   # ← замінити на власний префікс!

In [33]:

# 4. Три функции для работы с данными (из условия)
def read_json(name):
    raw = source.download_blob(name).readall()
    return pd.read_json(BytesIO(raw))

def read_parquet(path):
    raw = lake.download_blob(f"{STUDENT}/{path}").readall()
    return pd.read_parquet(BytesIO(raw))

def write_parquet(df, path):
    buf = BytesIO()
    df.to_parquet(buf, index=False)
    buf.seek(0)
    lake.upload_blob(name=f"{STUDENT}/{path}", data=buf.getvalue(), overwrite=True)

**Завдання 1.2. Прочитайте shops.json і products.json та виведіть кількість рядків у кожному. Очікувані значення: 5 і 12.**

In [34]:
# Читаємо файли з джерела за допомогою функції
shops_df = read_json('shops.json')
products_df = read_json('products.json')

# Виводимо кількість рядків у кожному DataFrame
print(f"Кількість рядків у shops.json: {len(shops_df)}")
print(f"Кількість рядків у products.json: {len(products_df)}")

Кількість рядків у shops.json: 5
Кількість рядків у products.json: 12


**Завдання 2. Шар Bronze — сирі дані як є**

**Завдання 2.1. Прочитайте три файли замовлень і об’єднайте їх в один DataFrame через pd.concat. Не змінюйте ні типів, ні значень. Виведіть кількість рядків — має бути 980.**

In [35]:
# --- ЗАВДАННЯ 2.1: Об'єднання замовлень для Bronze шару ---

# Читаємо три файли замовлень за допомогою генератора списків
parts = [read_json(f"orders_2026_0{m}.json") for m in (1, 2, 3)]

# Об'єднуємо їх в один DataFrame без зміни типів чи значень
bronze_orders = pd.concat(parts, ignore_index=True)

# Перевіряємо кількість рядків
print(f"Кількість рядків у bronze_orders: {len(bronze_orders)}")

Кількість рядків у bronze_orders: 980


**Завдання 2.2. Збережіть результат у bronze/orders.parquet.**

In [36]:
# --- ЗАВДАННЯ 2.2: Збереження замовлень у шар Bronze ---

# Записуємо об'єднаний DataFrame у форматі Parquet
write_parquet(bronze_orders, "bronze/orders.parquet")

# Перевірка: спробуємо прочитати файл назад з озера, щоб переконатися в успішному записі
check_bronze_orders = read_parquet("bronze/orders.parquet")
print(f"Файл успішно записано. Кількість рядків у збереженому файлі: {len(check_bronze_orders)}")

Файл успішно записано. Кількість рядків у збереженому файлі: 980


**Завдання 2.3. Збережіть довідник кав’ярень у bronze/shops.parquet, а меню — у bronze/products.parquet, теж без жодних змін.**

In [37]:
# --- ЗАВДАННЯ 2.3: Збереження довідників у шар Bronze ---

# Записуємо довідник кав'ярень
write_parquet(shops_df, "bronze/shops.parquet")

# Записуємо меню
write_parquet(products_df, "bronze/products.parquet")

# Перевірка запису (необов'язково, для самоконтролю)
print(f"Збережено кав'ярень: {len(read_parquet('bronze/shops.parquet'))}")
print(f"Збережено товарів меню: {len(read_parquet('bronze/products.parquet'))}")

Збережено кав'ярень: 5
Збережено товарів меню: 12


**Завдання 2.4. Виведіть перелік файлів у власному префіксі озера — має бути 3 файли:**

In [38]:
# Шукаємо файли конкретно в папці bronze для поточної роботи
for blob in lake.list_blobs(name_starts_with=f"{STUDENT}/bronze/"):
    print(blob.name, blob.size)

olkhovets_olena/bronze/orders.parquet 18620
olkhovets_olena/bronze/products.parquet 3609
olkhovets_olena/bronze/readings/readings.parquet 5070
olkhovets_olena/bronze/sensors/sensors.parquet 4125
olkhovets_olena/bronze/shops.parquet 4014
olkhovets_olena/bronze/stations/stations.parquet 4848


**Завдання 3. Шар Silver — очищені дані**

**Завдання 3.1.** Прочитайте bronze/orders.parquet і виведіть чотири показники якості. Очікувані значення наведені в дужках — звіряйтеся з ними:
кількість повних дублікатів рядків (20);
кількість унікальних значень shop_code (10);
кількість унікальних значень payment_type (4);
кількість порожніх значень qty (36).*

In [39]:
# --- ЗАВДАННЯ 3.1: Показники якості для orders ---

# Читаємо дані з шару Bronze
orders_silver_check = read_parquet("bronze/orders.parquet")

# 1. Кількість повних дублікатів рядків (очікувано: 20)
duplicates_count = orders_silver_check.duplicated().sum()

# 2. Кількість унікальних значень shop_code (очікувано: 10)
unique_shops = orders_silver_check['shop_code'].nunique()

# 3. Кількість унікальних значень payment_type (очікувано: 4)
unique_payments = orders_silver_check['payment_type'].nunique()

# 4. Кількість порожніх значень qty (очікувано: 36)
missing_qty = orders_silver_check['qty'].isna().sum()

# Виведення результатів
print(f"Кількість повних дублікатів: {duplicates_count}")
print(f"Кількість унікальних shop_code: {unique_shops}")
print(f"Кількість унікальних payment_type: {unique_payments}")
print(f"Кількість порожніх qty: {missing_qty}")

Кількість повних дублікатів: 20
Кількість унікальних shop_code: 10
Кількість унікальних payment_type: 4
Кількість порожніх qty: 36


**Завдання 3.2.** Приберіть повні дублікати рядків.*

In [40]:
# --- ЗАВДАННЯ 3.2: Видалення повних дублікатів рядків ---

# Очищуємо DataFrame від повних дублікатів
orders_no_duplicates = orders_silver_check.drop_duplicates(ignore_index=True)

# Перевірка: кількість рядків після видалення дублікатів (980 - 20 = 960)
print(f"Кількість рядків після видалення дублікатів: {len(orders_no_duplicates)}")

# Додаємо .copy() в кінці
orders_no_duplicates = orders_silver_check.drop_duplicates(ignore_index=True).copy()
print(f"Кількість рядків після видалення дублікатів: {len(orders_no_duplicates)}")


Кількість рядків після видалення дублікатів: 960
Кількість рядків після видалення дублікатів: 960


**Завдання 3.3.** Приведіть shop_code до єдиного вигляду: прибрати пробіли з країв і перевести у верхній регістр. Після цього унікальних значень має бути 5: S01, S02, S03, S04, S05*

In [41]:
# --- ЗАВДАННЯ 3.3: Очищення стовпчика shop_code ---

# Очищаємо пробіли та приводимо до верхнього регістру
orders_no_duplicates['shop_code'] = orders_no_duplicates['shop_code'].str.strip().str.upper()

# Перевірка: виводимо кількість унікальних значень та самі значення
unique_shops_silver = orders_no_duplicates['shop_code'].nunique()
unique_shops_list = sorted(orders_no_duplicates['shop_code'].unique())

print(f"Кількість унікальних shop_code: {unique_shops_silver}")
print(f"Список унікальних значень: {unique_shops_list}")

Кількість унікальних shop_code: 5
Список унікальних значень: ['S01', 'S02', 'S03', 'S04', 'S05']


**Завдання 3.4.** Приведіть payment_type до вигляду Card / Cash. Унікальних значень має стати 2.*

In [42]:
# --- ЗАВДАННЯ 3.4: Очищення стовпчика payment_type ---

# Спочатку приведемо до нижнього регістру та приберемо пробіли для надійності
orders_no_duplicates['payment_type'] = orders_no_duplicates['payment_type'].str.strip().str.lower()

# Замінюємо варіанти на уніфіковані "Card" або "Cash"
# (pandas підтримує заміну через словник, де ключ — старе значення, а значення — нове)
mapping = {
    'card': 'Card',
    'cash': 'Cash',
    # якщо у ваших даних є інші варіанти (наприклад, 'картка' або 'готівка'), додайте їх сюди
}
orders_no_duplicates['payment_type'] = orders_no_duplicates['payment_type'].replace(mapping)

# Перевірка: виводимо унікальні значення
unique_payments_silver = orders_no_duplicates['payment_type'].nunique()
unique_payments_list = orders_no_duplicates['payment_type'].unique()

print(f"Кількість унікальних payment_type: {unique_payments_silver}")
print(f"Список унікальних значень: {unique_payments_list}")

Кількість унікальних payment_type: 2
Список унікальних значень: ['Card' 'Cash']


**Завдання 3.5.** Замініть порожні qty на 1 і переведіть колонку в цілий тип. 

In [43]:
# --- ЗАВДАННЯ 3.5: Заповнення пропусків та зміна типу для qty ---

# Замінюємо порожні значення (NaN) на 1
orders_no_duplicates['qty'] = orders_no_duplicates['qty'].fillna(1)

# Переводимо стовпчик у цілий тип (int)
orders_no_duplicates['qty'] = orders_no_duplicates['qty'].astype(int)

# Перевірка: рахуємо кількість порожніх значень та перевіряємо тип даних стовпчика
missing_qty_after = orders_no_duplicates['qty'].isna().sum()
column_type = orders_no_duplicates['qty'].dtype

print(f"Залишилося порожніх значень у qty: {missing_qty_after}")
print(f"Тип даних стовпчика qty: {column_type}")

Залишилося порожніх значень у qty: 0
Тип даних стовпчика qty: int64


**Завдання 3.6.** Переведіть order_ts у тип дати й часу та додайте колонку order_date — тільки дата, без часу.*

In [44]:
# --- ЗАВДАННЯ 3.6: Приведение типов для order_ts и создание order_date ---

# Переводим order_ts в тип datetime
orders_no_duplicates['order_ts'] = pd.to_datetime(orders_no_duplicates['order_ts'])

# Добавляем колонку order_date (только дата, без времени)
orders_no_duplicates['order_date'] = orders_no_duplicates['order_ts'].dt.date

# Проверка: выводим типы данных и первые строки для самоконтроля
print(orders_no_duplicates[['order_ts', 'order_date']].dtypes)
print("\nПервые 3 строки:")
print(orders_no_duplicates[['order_ts', 'order_date']].head(3))

order_ts      datetime64[ns]
order_date            object
dtype: object

Первые 3 строки:
             order_ts  order_date
0 2026-01-27 12:05:00  2026-01-27
1 2026-01-03 09:30:00  2026-01-03
2 2026-01-30 11:35:00  2026-01-30


**Завдання 3.7.** Додайте колонку amount = qty × unit_price.*

In [45]:
# --- ЗАВДАННЯ 3.7: Расчет суммы заказа (amount) ---

# Добавляем колонку amount как произведение количества на цену за единицу
orders_no_duplicates['amount'] = orders_no_duplicates['qty'] * orders_no_duplicates['unit_price']

# Проверка: выводим первые 3 строки для контроля результата
print(orders_no_duplicates[['qty', 'unit_price', 'amount']].head(3))

   qty  unit_price  amount
0    2          80     160
1    2          80     160
2    1          95      95


**Завдання 3.8.** Збережіть результат у silver/orders.parquet.*

In [46]:
# --- ЗАВДАННЯ 3.8: Сохранение очищенных заказов в слой Silver ---

# Записываем обработанный DataFrame в слой Silver
write_parquet(orders_no_duplicates, "silver/orders.parquet")

# Проверка: считываем обратно и проверяем количество строк (должно быть 960)
silver_orders_check = read_parquet("silver/orders.parquet")
print(f"Файл успешно сохранен в Silver. Количество строк: {len(silver_orders_check)}")

Файл успешно сохранен в Silver. Количество строк: 960


**Завдання 3.9.** Перекладіть bronze/shops.parquet і bronze/products.parquet у silver/shops.parquet і silver/products.parquet без змін — ці два довідники уже чисті, але шар Silver має бути повним.

In [47]:
# --- ЗАВДАННЯ 3.9: Перенесення довідників у шар Silver ---

# Читаємо чисті довідники з шару Bronze
shops_bronze = read_parquet("bronze/shops.parquet")
products_bronze = read_parquet("bronze/products.parquet")

# Записуємо їх у шар Silver без змін
write_parquet(shops_bronze, "silver/shops.parquet")
write_parquet(products_bronze, "silver/products.parquet")

# Перевірка: виводимо кількість рядків для контролю
print(f"Довідник shops перенесено у Silver. Рядків: {len(read_parquet('silver/shops.parquet'))}")
print(f"Довідник products перенесено у Silver. Рядків: {len(read_parquet('silver/products.parquet'))}")

Довідник shops перенесено у Silver. Рядків: 5
Довідник products перенесено у Silver. Рядків: 12


**Завдання 4. Шар Gold — зоряна схема**

**Завдання 4.1.** Побудуйте dim_shop: візьміть silver/shops.parquet, відсортуйте за shop_code і додайте першою колонкою shop_key зі значеннями від 1 до 5.

In [48]:
# --- ЗАВДАННЯ 4.1: Побудова виміру dim_shop ---

# Читаємо довідник магазинів із шару Silver
silver_shops = read_parquet("silver/shops.parquet")

# Сортуємо за бізнес-ключем та скидаємо індекси
dim_shop = silver_shops.sort_values("shop_code").reset_index(drop=True)

# Додаємо першою колонкою сурогатний ключ shop_key (від 1 до 5)
dim_shop.insert(0, "shop_key", range(1, len(dim_shop) + 1))

# Перевірка: виводимо отриманий вимір
print("Таблиця виміру dim_shop:")
print(dim_shop)

Таблиця виміру dim_shop:
   shop_key shop_code     shop_name     city  district opened_date  seats
0         1       S01      Kava Hub     Kyiv     Podil  2019-04-15     24
1         2       S02         Zerno     Kyiv  Pechersk  2021-09-01     16
2         3       S03  Rynok Coffee     Lviv    Center  2018-06-20     30
3         4       S04         Bereg    Odesa   Arkadia  2022-05-10     40
4         5       S05         Steam  Kharkiv    Center  2023-02-01     12


**Завдання 4.2.** Так само побудуйте dim_product із product_key від 1 до 12, відсортувавши за product_code.

In [49]:
# --- ЗАВДАННЯ 4.2: Побудова виміру dim_product ---

# Читаємо довідник товарів із шару Silver
silver_products = read_parquet("silver/products.parquet")

# Сортуємо за бізнес-ключем та скидаємо індекси
dim_product = silver_products.sort_values("product_code").reset_index(drop=True)

# Додаємо першою колонкою сурогатний ключ product_key (від 1 до 12)
dim_product.insert(0, "product_key", range(1, len(dim_product) + 1))

# Перевірка: виводимо отриманий вимір для контролю
print("Таблиця виміру dim_product:")
print(dim_product)

Таблиця виміру dim_product:
    product_key product_code  product_name category size  price
0             1          P01      Espresso   Coffee    S     45
1             2          P02     Americano   Coffee    M     55
2             3          P03    Cappuccino   Coffee    M     70
3             4          P04         Latte   Coffee    L     80
4             5          P05    Flat White   Coffee    M     75
5             6          P06     Green Tea      Tea    M     50
6             7          P07    Herbal Tea      Tea    M     50
7             8          P08     Croissant   Bakery    -     60
8             9          P09  Cinnamon Bun   Bakery    -     65
9            10          P10      Sandwich   Bakery    -    110
10           11          P11    Cheesecake  Dessert    -     95
11           12          P12       Brownie  Dessert    -     85


**Завдання 4.3.** Побудуйте fact_sales: приєднайте до silver/orders.parquet обидва виміри й замініть коди на ключі.

In [50]:
# --- ЗАВДАННЯ 4.3: Побудова таблиці фактів fact_sales ---

# 1. Читаємо замовлення з шару Silver
silver_orders = read_parquet("silver/orders.parquet")

# 2. Приєднуємо dim_shop за бізнес-ключем 'shop_code'
fact_sales = pd.merge(silver_orders, dim_shop[['shop_code', 'shop_key']], on='shop_code', how='left')

# 3. Приєднуємо dim_product за бізнес-ключем 'product_code'
fact_sales = pd.merge(fact_sales, dim_product[['product_code', 'product_key']], on='product_code', how='left')

# 4. Відбираємо колонки і виставляємо їх у заданому порядку
target_columns = [
    'order_id', 'shop_key', 'product_key', 'order_date', 
    'qty', 'unit_price', 'amount', 'payment_type'
]
fact_sales = fact_sales[target_columns]

# Перевірка: виводимо структуру та кількість рядків (має залишитися 960)
print(f"Кількість рядків у fact_sales: {len(fact_sales)}")
print("\nПерші 3 строки таблиці фактів:")
print(fact_sales.head(3))

Кількість рядків у fact_sales: 960

Перші 3 строки таблиці фактів:
  order_id  shop_key  product_key  order_date  qty  unit_price  amount  \
0  O500259         4            4  2026-01-27    2          80     160   
1  O500252         5            4  2026-01-03    2          80     160   
2  O500257         4           11  2026-01-30    1          95      95   

  payment_type  
0         Card  
1         Card  
2         Cash  


**Завдання 4.4.** Побудуйте другу таблицю фактів fact_daily_sales: згрупуйте fact_sales за shop_key і order_date.

In [51]:
# --- ЗАВДАННЯ 4.4: Построение агрегированной таблицы фактів fact_daily_sales ---

# Группируем данные и рассчитываем нужные метрики
fact_daily_sales = fact_sales.groupby(['shop_key', 'order_date']).agg(
    orders_count=('order_id', 'nunique'),  # Количество уникальных заказов
    items_sold=('qty', 'sum'),             # Сумма проданных товаров
    revenue=('amount', 'sum')              # Сумма выручки
).reset_index()

# Проверка: выводим первые 5 строк для контроля структуры
print("Таблица агрегированных фактов fact_daily_sales:")
print(fact_daily_sales.head())
print(f"\nОбщее количество строк в агрегированной таблице: {len(fact_daily_sales)}")

Таблица агрегированных фактов fact_daily_sales:
   shop_key  order_date  orders_count  items_sold  revenue
0         1  2026-01-01             2           3      235
1         1  2026-01-02             1           1       95
2         1  2026-01-03             6           9      785
3         1  2026-01-04             2           3      225
4         1  2026-01-05             2           3      170

Общее количество строк в агрегированной таблице: 396


**Завдання 4.5.** Збережіть усі чотири таблиці: gold/dim_shop.parquet, gold/dim_product.parquet, gold/fact_sales.parquet, gold/fact_daily_sales.parquet.

In [52]:
# --- ЗАВДАННЯ 4.5: Збереження таблиць у шар Gold та валідація ---

# 1. Зберігаємо виміри та таблиці фактів
write_parquet(dim_shop, "gold/dim_shop.parquet")
write_parquet(dim_product, "gold/dim_product.parquet")
write_parquet(fact_sales, "gold/fact_sales.parquet")
write_parquet(fact_daily_sales, "gold/fact_daily_sales.parquet")

# 2. Валідація: перевірка сум виторгу
total_amount_fact = fact_sales['amount'].sum()
total_revenue_daily = fact_daily_sales['revenue'].sum()

print("--- ПЕРЕВІРКА ВАЛІДНОСТІ ДАНИХ ---")
print(f"Загальний виторг у fact_sales:       {total_amount_fact:.2f}")
print(f"Загальний виторг у fact_daily_sales: {total_revenue_daily:.2f}")

if round(total_amount_fact, 2) == round(total_revenue_daily, 2):
    print("✅ Успіх! Суми збігаються. Помилок у групуванні немає.")
else:
    print("❌ Помилка! Суми розбігаються. Перевірте логіку групування в Завданні 4.4.")

--- ПЕРЕВІРКА ВАЛІДНОСТІ ДАНИХ ---
Загальний виторг у fact_sales:       90320.00
Загальний виторг у fact_daily_sales: 90320.00
✅ Успіх! Суми збігаються. Помилок у групуванні немає.


**Завдання 5. Перевірка та аналітика**

**Завдання 5.1.** Виведіть кількість рядків у кожній з чотирьох таблиць Gold. Очікувано: 5, 12, 960, 396.

In [53]:
# --- ЗАВДАННЯ 5.1: Перевірка кількості рядків у таблицях Gold ---

# Читаємо таблиці з шару Gold
gold_dim_shop = read_parquet("gold/dim_shop.parquet")
gold_dim_product = read_parquet("gold/dim_product.parquet")
gold_fact_sales = read_parquet("gold/fact_sales.parquet")
gold_fact_daily_sales = read_parquet("gold/fact_daily_sales.parquet")

# Виводимо кількість рядків у кожній таблиці
print(f"Кількість рядків у dim_shop:        {len(gold_dim_shop)}")
print(f"Кількість рядків у dim_product:     {len(gold_dim_product)}")
print(f"Кількість рядків у fact_sales:       {len(gold_fact_sales)}")
print(f"Кількість рядків у fact_daily_sales: {len(gold_fact_daily_sales)}")

Кількість рядків у dim_shop:        5
Кількість рядків у dim_product:     12
Кількість рядків у fact_sales:       960
Кількість рядків у fact_daily_sales: 396


**Завдання 5.2.** Перевірте, що всі shop_key з fact_sales присутні в dim_shop, а всі product_key — в dim_product. Обидві перевірки мають вивести True.

In [54]:
# --- ЗАВДАННЯ 5.2: Перевірка цілісності зв'язків (Foreign Keys) ---

# Перевіряємо, чи всі магазини з таблиці фактів є у вимірі dim_shop
check_shops = gold_fact_sales["shop_key"].isin(gold_dim_shop["shop_key"]).all()

# Перевіряємо, чи всі товари з таблиці фактів є у вимірі dim_product
check_products = gold_fact_sales["product_key"].isin(gold_dim_product["product_key"]).all()

# Виводимо результати перевірки
print(f"Усі shop_key присутні в dim_shop: {check_shops}")
print(f"Усі product_key присутні в dim_product: {check_products}")

Усі shop_key присутні в dim_shop: True
Усі product_key присутні в dim_product: True


**Завдання 5.3.** Перевірте, що сума amount у fact_sales дорівнює сумі revenue у fact_daily_sales. Має вивести True.

In [55]:
# --- ЗАВДАННЯ 5.3: Перевірка рівності сум виторгу ---

# Рахуємо суми в обох таблицях
sum_fact_sales = gold_fact_sales["amount"].sum()
sum_daily_sales = gold_fact_daily_sales["revenue"].sum()

# Порівнюємо округлені до копійок значення
is_equal = round(sum_fact_sales, 2) == round(sum_daily_sales, 2)

# Виводимо результат перевірки
print(is_equal)

True


**Завдання 5.4.** Порахуйте виторг за категоріями меню: приєднайте dim_product до fact_sales, згрупуйте за category, відсортуйте за спаданням суми amount. 

In [56]:
# --- ЗАВДАННЯ 5.4: Розрахунок виторгу за категоріями меню ---

# 1. Приєднуємо dim_product до fact_sales за ключем product_key
merged_sales = pd.merge(gold_fact_sales, gold_dim_product[['product_key', 'category']], on='product_key', how='left')

# 2. Групуємо за категорією, рахуємо суму amount та перейменовуємо стовпчик у 'виторг'
category_revenue = merged_sales.groupby('category')['amount'].sum().reset_index()
category_revenue = category_revenue.rename(columns={'amount': 'виторг'})

# 3. Сортуємо за спаданням суми виторгу
category_revenue = category_revenue.sort_values(by='виторг', ascending=False).reset_index(drop=True)

# Виведення результату
print(category_revenue)

  category  виторг
0   Coffee   49915
1   Bakery   19955
2  Dessert   14850
3      Tea    5600


**Завдання 5.5.** Порахуйте виторг за кав’ярнями: приєднайте dim_shop до fact_sales, згрупуйте за shop_name, відсортуйте за спаданням. Перше місце — Rynok Coffee з виторгом 20075.

In [57]:
# --- ЗАВДАННЯ 5.5: Розрахунок виторгу за кав’ярнями ---

# 1. Приєднуємо dim_shop до fact_sales за ключем shop_key
merged_shops = pd.merge(gold_fact_sales, gold_dim_shop[['shop_key', 'shop_name']], on='shop_key', how='left')

# 2. Групуємо за назвою, рахуємо суму amount та перейменовуємо стовпчик у 'виторг'
shop_revenue = merged_shops.groupby('shop_name')['amount'].sum().reset_index()
shop_revenue = shop_revenue.rename(columns={'amount': 'виторг'})

# 3. Сортуємо за спаданням суми виторгу
shop_revenue = shop_revenue.sort_values(by='виторг', ascending=False).reset_index(drop=True)

# Виведення результату
print(shop_revenue)

      shop_name  виторг
0  Rynok Coffee   20075
1      Kava Hub   19290
2         Steam   18275
3         Bereg   16750
4         Zerno   15930


In [58]:
# --- ФІНАЛЬНА ПЕРЕВІРКА ТЕХНІЧНИХ ВИМОГ ---

# Список усіх цільових файлів для цієї роботи
target_files = [
    "bronze/orders.parquet", "bronze/shops.parquet", "bronze/products.parquet",
    "silver/orders.parquet", "silver/shops.parquet", "silver/products.parquet",
    "gold/dim_shop.parquet", "gold/dim_product.parquet", "gold/fact_sales.parquet", "gold/fact_daily_sales.parquet"
]

print("=== Перевірка наявності файлів у вашому префіксі ===")
found_count = 0

for file_path in target_files:
    full_name = f"{STUDENT}/{file_path}"
    # Перевіряємо наявність файлу в озері
    blobs = list(lake.list_blobs(name_starts_with=full_name))
    
    if blobs:
        print(f"✅ Знайдено ({blobs[0].size} байт): {file_path}")
        found_count += 1
    else:
        print(f"❌ ВІДСУТНІЙ ФАЙЛ: {file_path}")

print("\n=== Підсумок перевірки ===")
print(f"Усього знайдено файлів для кав'ярень: {found_count} із 10")

if found_count == 10:
    print("🎉 Вітаємо! Усі 10 файлів на місці. Технічні вимоги виконано!")
else:
    print("⚠️ Чекаут не пройдено. Перевірте, які файли позначені хрестиком (❌) та перезапустіть відповідні комірки.")

=== Перевірка наявності файлів у вашому префіксі ===
✅ Знайдено (18620 байт): bronze/orders.parquet
✅ Знайдено (4014 байт): bronze/shops.parquet
✅ Знайдено (3609 байт): bronze/products.parquet
✅ Знайдено (22915 байт): silver/orders.parquet
✅ Знайдено (4014 байт): silver/shops.parquet
✅ Знайдено (3609 байт): silver/products.parquet
✅ Знайдено (4600 байт): gold/dim_shop.parquet
✅ Знайдено (4241 байт): gold/dim_product.parquet
✅ Знайдено (13765 байт): gold/fact_sales.parquet
✅ Знайдено (5501 байт): gold/fact_daily_sales.parquet

=== Підсумок перевірки ===
Усього знайдено файлів для кав'ярень: 10 із 10
🎉 Вітаємо! Усі 10 файлів на місці. Технічні вимоги виконано!


In [59]:
# --- ФІНАЛЬНА ПЕРЕВІРКА З ПОВНИМИ ШЛЯХАМИ ---

target_files = [
    "bronze/orders.parquet", "bronze/shops.parquet", "bronze/products.parquet",
    "silver/orders.parquet", "silver/shops.parquet", "silver/products.parquet",
    "gold/dim_shop.parquet", "gold/dim_product.parquet", "gold/fact_sales.parquet", "gold/fact_daily_sales.parquet"
]

print("=== Повні шляхи до файлів в озері ===")
for file_path in target_files:
    full_name = f"{STUDENT}/{file_path}"
    blobs = list(lake.list_blobs(name_starts_with=full_name))
    
    if blobs:
        print(f"✅ {blobs[0].name}") # Виведе повне ім'я з префіксом olkhovets_olena

=== Повні шляхи до файлів в озері ===
✅ olkhovets_olena/bronze/orders.parquet
✅ olkhovets_olena/bronze/shops.parquet
✅ olkhovets_olena/bronze/products.parquet
✅ olkhovets_olena/silver/orders.parquet
✅ olkhovets_olena/silver/shops.parquet
✅ olkhovets_olena/silver/products.parquet
✅ olkhovets_olena/gold/dim_shop.parquet
✅ olkhovets_olena/gold/dim_product.parquet
✅ olkhovets_olena/gold/fact_sales.parquet
✅ olkhovets_olena/gold/fact_daily_sales.parquet
